### ISSUE TITLE
Option to disable implicitly adding `using` directives

### ISSUE THREAD

**Issue by logiclrd** [dotnet/roslyn#74492]
If I type the name of an item from a namespace that is not imported, it automatically -- silently -- adds a `using` directive to the top of the file. Is it possible to turn this off? I don't want to get rid of the IntelliSense listing types from other namespaces, that's useful, but I personally would prefer selecting a non-imported type to simply generate a compile-time error. Many, many times now, I've found bizarre `using` directives at the top of files because at some point, I fat-fingered a type name. I backspaced it and corrected it, but the unnecessary `using` directive, added out of view, remains.

**Comment by MisutaaAsriel** [comment]
This actually causes issues working with Udon and the VRChat SDK. Code will frequently add namespaces that are unavailable to Udon, or namespaces **that are not even in use**, just because of autocomplete suggestions.

Like, genuinely, the namespaces are not in use.

I frequently end up with the `UnityEditor` namespace directive being added, for example, that I do not want. Same with various `System` related namespace directives. But still being able to *reference* functions from outside namespaces is pretty important, as I may not know exactly which namespace a function lives in.

This is especially a problem for me because U# is a subset of C# with whitelisted namespaces and functions. If it references a namespace that is not available, outside of a platform definition that limits compilation to the editor, the compile fails.

My expectation is that only functions I *use* should have directives added, and only if I **choose them** from the autocomplete suggestion.

**Comment by CyrusNajmabadi** [comment]
@MisutaaAsriel Just turn off "Show Items from Unimported Namespaces".

![Image](https://github.com/user-attachments/assets/a4131d0f-c38c-44cc-b30d-90bd8a9abd74)

**Comment by CyrusNajmabadi** [comment]
The option to disable this exists.  So closing this out.

**Comment by MisutaaAsriel** [comment]
> @MisutaaAsriel Just turn off "Show Items from Unimported Namespaces".
> 
> ![Image](https://github.com/user-attachments/assets/a4131d0f-c38c-44cc-b30d-90bd8a9abd74)

@CyrusNajmabadi that literally does not solve the issue and is the complete opposite of what both OP & I mentioned an issue with.

You have incorrectly closed this issue.

We _**WANT**_ to be shown items from unimported namespaces **in intellisense suggestions**.

The **problem** is that it sometimes imports namespaces automatically when going through the results or typing similarly named functions.

In other words, the option to automatically import namespaces should have the option to only import *explicitly* when *selected* from an intellisence suggestion.

If I _select and complete_ a function listed in another namespace from suggestions, I want it to import the namespace.

If I scroll through the functions in intellisense suggestions, or if type a similarly named function, I _don't_ want it to import the namespace.

This is **a very important distinction**.

**Comment by CyrusNajmabadi** [comment]
There's no capability in Roslyn that I know of that causes that.  If you can provide a repro, we can fix, or find the team causing this.

**Comment by CyrusNajmabadi** [comment]
@MisutaaAsriel the op says:

> If I type the name of an item from a namespace that is not imported, it automatically -- silently -- adds a using directive to the top of the file. 

But you say you want that behavior. 

> In other words, the option to automatically import namespaces should have the option to only import explicitly when selected from an intellisence suggestion.

This is how it works. Only the items in completion that are pulled in from unimported namespaces will add an import.

**Comment by CyrusNajmabadi** [comment]
@MisutaaAsriel 

> If I scroll through the functions in intellisense suggestions, or if type a similarly named function, I don't want it to import the namespace.

That should be how it works. Can you supply a repro where it does import the namespace when typing a similarly named function?

**Comment by MisutaaAsriel** [comment]
> > If I type the name of an item from a namespace that is not imported, it automatically -- silently -- adds a using directive to the top of the file.
> 
> But you say you want that behavior.

No. I want *suggestions* from other namespaces, but I only want it to *import* if I explicitly select an item from intellisense. If I type the full function, or merely browse the list of suggestions with my arrow keys, I do not.

> > If I scroll through the functions in intellisense suggestions, or if type a similarly named function, I don't want it to import the namespace.
> 
> That should be how it works. Can you supply a repro where it does import the namespace when typing a similarly named function?

Not currently. But I can say it happens at least once in any Unity project.

And to clarify, I should have said *identically* named functions/classes/variables/etc.

So if `namespace.a` and `namespace.b` both have, say, a class named `SomeClass`, intellisense sometimes imports the first namespace it finds with a `using` directive when typing out, say, `SomeClass obj;`.

**Comment by CyrusNajmabadi** [comment]
> but I only want it to import if I explicitly select an item from intellisense.

That's how it works.  Only if the item is commited from intellisense do we add the import.  The item in intellisense is the only thing that even has the info to add such an import.  So if you're typing something else, or selecting anything else, those won't have hte info to even add the import in the first place.

> Not currently.

If you do, please open a new issue with the repro.  Thanks.
